### PR TITLE
Improve blurred content of embedded windows

### LINKED ISSUE #54030: Popups (OptionButton, tooltips) placed and scaled wrong when using `canvas_items` stretch mode or scale factor different from `1.0` (DisplayServer regression)
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

### Godot version

4.0-dev.20211004_win64

### System information

Windows 10

### Issue description

When using OptionButton and choosing **canvas_items** mode in Project Settings->Display, popup of OptionButton is placed and scaled wrong (fonts too). When using "disabled" mode it is placed correctly.

popupmenu_issue1

I originally thought that this is related to my custom size of display, but it seems only "canvas_items" does this.

popupmenu_issue2

I've added project, so you can test. but it's pretty obvious, I have no scaling or complex settings on nodes.

### Steps to reproduce

Set canvas_items mode and use PopupButton.

### Minimal reproduction project

[game-client_issue_popup_menu.zip](https://github.com/godotengine/godot/files/7382574/game-client_issue_popup_menu.zip)

### PR DESCRIPTION
**Pull request by scgm0**
Helps: #54030
Helps: #95387

I honestly don't know if I'm doing it the right way, but the issue seems to be forgotten and after a flurry of attempts I'm getting good results.
before:
图片
after:
图片
example:
[embedded-window.zip](https://github.com/user-attachments/files/17233855/embedded-window.zip)

### CODE REVIEW

**Comment by hsvfan-jan** [comment]
In the screenshot, the circles of Test2 and Test3 seem to have become an octagon (the newly introduced white outline probably making it a full circle I'd assume)

**Comment by scgm0** [comment]
> In the screenshot, the circles of Test2 and Test3 seem to have become an octagon (the newly introduced white outline probably making it a full circle I'd assume)

It seems to be a problem caused by `viewport_set_global_canvas_transform`. I don’t know where the specific problem is.

**Comment by scgm0** [comment]
> In the screenshot, the circles of Test2 and Test3 seem to have become an octagon (the newly introduced white outline probably making it a full circle I'd assume)

When I manually assign the `radio_unchecked` icon in the theme (`radio_unchecked.svg` was copied from godot), there is no white border around the icon, so maybe the problem is related to the theme cache?
图片
图片

**Comment by scgm0** [comment]
The problem of `viewport_set_global_canvas_transform` should not be within the scope of this PR. What needs to be discussed now is whether the solution of this PR is qualified and whether there are any side effects.

**Comment by Summersay415** [comment]
Tested locally, works fine. It doesn't remove all blurriness, but decreases it a lot.

**Comment by scgm0** [comment]
> Tested locally, works fine. It doesn't remove all blurriness, but decreases it a lot. Before: изображение After: изображение

Have you enabled the font msdf?

**Comment by Summersay415** [comment]
No

**Comment by scgm0** [comment]
> No

Give it a try?

**Comment by eldidou** [comment]
Hello, It seems that this fix solves https://github.com/godotengine/godot/issues/95387, with no side-effect seen so far.

**Comment by scgm0** [comment]
> Hello, It seems that this fix solves #95387, with no side-effect seen so far.

That's great if there's no problem.

**Comment by timothyqiu** [comment]
Tooltip and popup menu are still blurred: [test-4.zip](https://github.com/user-attachments/files/17536316/test-4.zip)

Peek 2024-10-28 09-45

**Comment by scgm0** [comment]
> Tooltip and popup menu are still blurred: [test-4.zip](https://github.com/user-attachments/files/17536316/test-4.zip)
> 
> Peek 2024-10-28 09-51 [ Peek 2024-10-28 09-51 ](https://private-user-images.githubusercontent.com/372476/380552056-b0d0389d-0735-4f62-93e6-e9105c590fd7.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAwODA2NDEsIm5iZiI6MTczMDA4MDM0MSwicGF0aCI6Ii8zNzI0NzYvMzgwNTUyMDU2LWIwZDAzODlkLTA3MzUtNGY2Mi05M2U2LWU5MTA1YzU5MGZkNy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyOFQwMTUyMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wOTdiMGU3NjY0ZWZlZGRjODU2YzVlMDAwZjI5ZTFmMzFjN2QwN2M3MGNjODI1MTNjZDA4MDAwMmE3NWRlOGVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bzefjGo70wf3h-LqMVwlXiGBtfpFhOs7JaDq3yGoOAA) [ ](https://private-user-images.githubusercontent.com/372476/380552056-b0d0389d-0735-4f62-93e6-e9105c590fd7.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAwODA2NDEsIm5iZiI6MTczMDA4MDM0MSwicGF0aCI6Ii8zNzI0NzYvMzgwNTUyMDU2LWIwZDAzODlkLTA3MzUtNGY2Mi05M2U2LWU5MTA1YzU5MGZkNy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyOFQwMTUyMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wOTdiMGU3NjY0ZWZlZGRjODU2YzVlMDAwZjI5ZTFmMzFjN2QwN2M3MGNjODI1MTNjZDA4MDAwMmE3NWRlOGVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bzefjGo70wf3h-LqMVwlXiGBtfpFhOs7JaDq3yGoOAA)
> 
> Texture filtering is not inherited I think.

What is blurred is only the texture with constant resolution, which does not belong to the content of this PR, Whether the texture filtering mode is inherited should be discussed separately?

**Comment by scgm0** [comment]
> Tooltip and popup menu are still blurred: [test-4.zip](https://github.com/user-attachments/files/17536316/test-4.zip)
> 
> Peek 2024-10-28 09-51 [ Peek 2024-10-28 09-51 ](https://private-user-images.githubusercontent.com/372476/380552056-b0d0389d-0735-4f62-93e6-e9105c590fd7.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAwODA2NDEsIm5iZiI6MTczMDA4MDM0MSwicGF0aCI6Ii8zNzI0NzYvMzgwNTUyMDU2LWIwZDAzODlkLTA3MzUtNGY2Mi05M2U2LWU5MTA1YzU5MGZkNy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyOFQwMTUyMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wOTdiMGU3NjY0ZWZlZGRjODU2YzVlMDAwZjI5ZTFmMzFjN2QwN2M3MGNjODI1MTNjZDA4MDAwMmE3NWRlOGVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bzefjGo70wf3h-LqMVwlXiGBtfpFhOs7JaDq3yGoOAA) [ ](https://private-user-images.githubusercontent.com/372476/380552056-b0d0389d-0735-4f62-93e6-e9105c590fd7.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAwODA2NDEsIm5iZiI6MTczMDA4MDM0MSwicGF0aCI6Ii8zNzI0NzYvMzgwNTUyMDU2LWIwZDAzODlkLTA3MzUtNGY2Mi05M2U2LWU5MTA1YzU5MGZkNy5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyOFQwMTUyMjFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wOTdiMGU3NjY0ZWZlZGRjODU2YzVlMDAwZjI5ZTFmMzFjN2QwN2M3MGNjODI1MTNjZDA4MDAwMmE3NWRlOGVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bzefjGo70wf3h-LqMVwlXiGBtfpFhOs7JaDq3yGoOAA)
> 
> Texture filtering is not inherited I think.

图片
图片
It still looks like this problem can be solved by not using the original texture, but a `CanvasTexture`, and setting the texture filtering mode in the `CanvasTexture`.
It makes me miss `3.x` being able to set the filter mode directly on the texture without the need for `CanvasTexture`. This is also an obvious change since `4.0`. How to improve it is beyond the scope of this PR.

**Review by timothyqiu — COMMENTED** [review]
In general, I think the blurring issues mentioned in #54030 and #95387 are caused by a combination of issues such as the window's scaling mode, resolution, texture filtering mode inheritance, and so on.

This PR resolves some of those but not all. So I'd suggest making the title a bit more detailed and not marking it as closing those two issues.

**Comment by scgm0** [comment]
> In general, I think the blurring issues mentioned in #54030 and #95387 are caused by a combination of issues such as the window's scaling mode, resolution, texture filtering mode inheritance, and so on.
> 
> This PR resolves some of those but not all. So I'd suggest making the title a bit more detailed and not marking it as closing those two issues.

So how should I change the title and issue markup?

**Comment by clayjohn** [comment]
I edited the title and removed the auto-close. 

@timothyqiu Do you think it is fine to go ahead with merging this as-is?

**Review by timothyqiu — APPROVED** [review]
Tested locally and it greatly reduced blurriness :+1:

**Comment by scgm0** [comment]
Just wondering if there is something wrong with having this pr so it hasn't been merged yet?

**Comment by akien-mga** [comment]
> Just wondering if there is something wrong with having this pr so it hasn't been merged yet?

There's nothing wrong a priori, but it needs reviews from maintainers familiar with the Viewport and Window code (such as @bruvzg or @Sauermann).

**Comment by bruvzg** [comment]
Code seems fine, will fully check it tomorrow.

**Review by bruvzg — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
